### PR TITLE
Implements Stylesheet.ownerNode

### LINKED ISSUE #23082: Implement StyleSheet.ownerNode
As @emilio mentioned in https://github.com/servo/servo/pull/23073#pullrequestreview-217996814 , we don't support `ownerNode` yet.

Spec: https://drafts.csswg.org/cssom/#dom-stylesheet-ownernode

https://github.com/servo/servo/blob/0ff87c8137030ece065f283552c28f4837cef5f3/components/script/dom/webidls/StyleSheet.webidl#L11

We need to uncomment it and implement it in `components/script/dom/stylesheet.rs` (basically just need to reuse the method added in #23073 ).

(But somehow I'm confused that the spec said it's an `Element or ProcessingInstruction` but we only have `Element` in `CSSStyleSheet`. Maybe we can only always return an `Element` for now? 🤔)

### PR DESCRIPTION
**Pull request by ghostd**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #23082 (GitHub issue number if applicable)

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by jdm** [comment]
At least ttwf-cssom-doc-ext-load-count.html has a pass because https://github.com/web-platform-tests/wpt/blob/50130dfa455612be36f0f861c880b384b67cdc7e/css/cssom/ttwf-cssom-doc-ext-load-count.html#L34 no longer throws an exception, which allows the next test to run and it now fails. I suspect that's the same reason for many of the new failures, but we should verify each of them.

**Comment by ghostd** [comment]
I'll check that this weekend

**Comment by ghostd** [comment]
Bug analysis:
* /css/cssom/MediaList2.xhtml: the missing 'ownerNode' property prevented to actually run the test (so not a regression)
* /css/cssom/ttwf-cssom-doc-ext-load-count.html:
  * unexpected pass, so it's ok
  * the "fail" was a false positive (it should never have been ok), it seems the <link> dynamically created is not added to the document.stylesheets collection (so not a regression, but an already existing bug)
* all /css/selectors/parsing/*html: one regression i'm going to fix

**Comment by ghostd** [comment]
The regression should be fixed, but i need help for the code modified in cssrule.list. I created a big nasty because my rust skills are too low an ownership issue. I tried somehting like this:
```rust
let loader = owner.map(|element| StylesheetLoader::for_element(&*element));
```
but i get this error:
```
error[E0515]: cannot return value referencing function parameter `element`
   --> components/script/dom/cssrulelist.rs:111:42
    |
111 |         let loader = owner.map(|element| StylesheetLoader::for_element(&*element));
    |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-------^
    |                                          |                               |
    |                                          |                               `element` is borrowed here
    |                                          returns a value referencing data owned by the current function
```
I understand the issue, but cannot get a fix :-/

Any advices are welcome :)

**Comment by CYBAI** [comment]
> Any advices are welcome :)

You might want to do it [like these lines](https://github.com/servo/servo/pull/23096/files#diff-38e2b0c88ad25af6e2f5c6e661724c41R110-R113) :)

**Comment by ghostd** [comment]
I tried but i didnt get a compiling code neither :-/ I'll try again this weekend after some sleep :D

**Comment by jdm** [comment]
I believe that `owner.as_ref().map(|element| StylesheetLoader::for_element(&**element));` might work.

**Comment by ghostd** [comment]
It works, thanks!!!

**Comment by jdm** [comment]
```
Diff in /repo/components/script/dom/cssrulelist.rs at line 109:
             .get_owner()
             .map(DomRoot::downcast::<HTMLElement>)
             .flatten();
-        let loader = owner.as_ref().map(|element| StylesheetLoader::for_element(&**element));
(B+        let loader = owner
(B+            .as_ref()
(B+            .map(|element| StylesheetLoader::for_element(&**element));
(B         let new_rule = css_rules.with_raw_offset_arc(|arc| {
             arc.insert_rule(
                 &parent_stylesheet.shared_lock,
Run `./mach fmt` to fix the formatting
```

**Comment by ghostd** [comment]
Fixed

File: components/script/dom/cssstyledeclaration.rs [hunk 1]
```diff
@@ -113,8 +113,15 @@ impl CSSStyleOwner {
...
                     // CSSStyleRule::SetSelectorText, which does the same thing.
-                    stylesheets_owner_from_node(rule.parent_stylesheet().owner().upcast::<Node>())
+                    if rule.parent_stylesheet().get_owner().is_some() {
+                        stylesheets_owner_from_node(
+                            rule.parent_stylesheet()
+                                .get_owner()
+                                .unwrap()
+                                .upcast::<Node>(),
+                        )
                         .invalidate_stylesheets();
```
Comment (jdm) [inline components/script/dom/cssstyledeclaration.rs:123]: Let's do this instead:
```rust
if let Some(owner) = rule.parent_stylesheet().get_owner() {
    stylesheets_owner_from_node(owner.upcast::<Node>())
        .invalidate_stylesheets();
}
```

File: components/script/dom/cssstylerule.rs [hunk 2]
```diff
@@ -117,8 +117,16 @@ impl CSSStyleRuleMethods for CSSStyleRule {
             let mut guard = self.cssrule.shared_lock().write();
             let stylerule = self.stylerule.write_with(&mut guard);
             mem::swap(&mut stylerule.selectors, &mut s);
-            stylesheets_owner_from_node(self.cssrule.parent_stylesheet().owner().upcast::<Node>())
+            if self.cssrule.parent_stylesheet().get_owner().is_some() {
+                stylesheets_owner_from_node(
```
Comment (jdm) [inline components/script/dom/cssstylerule.rs:121]: Same as previous comment: let's use `if let` instead.

File: components/script/dom/stylesheet.rs [hunk 3]
```diff
@@ -44,6 +46,11 @@ impl StyleSheetMethods for StyleSheet {
         self.href.clone()
     }
 
+    // https://drafts.csswg.org/cssom/#dom-stylesheet-ownernode
+    fn GetOwnerNode(&self) -> Option<DomRoot<Element>> {
+        self.downcast::<CSSStyleSheet>().unwrap().get_owner()
```
Comment (jdm) [inline components/script/dom/stylesheet.rs:51]: Might as well use `and_then` instead of unwrapping here.

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good!

File: components/script/dom/stylesheet.rs (hunk 3 above)
Comment (ghostd) [reply components/script/dom/stylesheet.rs]: Fixed

File: components/script/dom/cssstylerule.rs (hunk 2 above)
Comment (ghostd) [reply components/script/dom/cssstylerule.rs]: Fixed

File: components/script/dom/cssstyledeclaration.rs (hunk 1 above)
Comment (ghostd) [reply components/script/dom/cssstyledeclaration.rs]: Fixed (more readable, thanks)

**Comment by jdm** [comment]
@bors-servo r+
Thanks!

**Comment by CYBAI** [comment]
Need to update these test expectations for layout-2020 as well!

```
  â–¶ Unexpected subtest result in /css/cssom/ttwf-cssom-doc-ext-load-count.html:
  â”” PASS [expected FAIL] stylesheet.css should be unloaded and styleSheets.length === 0

  â–¶ Unexpected subtest result in /css/cssom/ttwf-cssom-doc-ext-load-count.html:
  â”‚ FAIL [expected PASS] stylesheet-1.css should be loaded and styleSheets.length === 1
  â”‚   â†’ assert_equals: styleSheets.length is incorrect: expected 1 but got 0
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/cssom/ttwf-cssom-doc-ext-load-count.html:55:22
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”” @http://web-platform.test:8000/css/cssom/ttwf-cssom-doc-ext-load-count.html:42:11


  â–¶ Unexpected subtest result in /css/cssom/MediaList2.xhtml:
  â”” PASS [expected FAIL] MediaList

  â–¶ Unexpected subtest result in /css/cssom/MediaList2.xhtml:
  â”‚ FAIL [expected PASS] MediaList.length
  â”‚   â†’ media is undefined
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:13:5
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:12:7
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”” @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:2:5

  â–¶ Unexpected subtest result in /css/cssom/MediaList2.xhtml:
  â”‚ FAIL [expected PASS] MediaList.item
  â”‚   â†’ media is undefined
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:24:5
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:23:7
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”” @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:2:5

  â–¶ Unexpected subtest result in /css/cssom/MediaList2.xhtml:
  â”‚ FAIL [expected PASS] MediaList getter
  â”‚   â†’ media is undefined
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:17:5
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:16:7
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”” @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:2:5

  â–¶ Unexpected subtest result in /css/cssom/MediaList2.xhtml:
  â”‚ FAIL [expected PASS] MediaList.mediaText
  â”‚   â†’ media is undefined
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:9:5
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”‚ @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:7:7
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:535:30
  â”” @http://web-platform.test:8000/css/cssom/MediaList2.xhtml:2:5


  â–¶ Unexpected subtest result in /css/cssom/idlharness.html:
  â”” PASS [expected FAIL] StyleSheet interface: attribute ownerNode

  â–¶ Unexpected subtest result in /css/cssom/idlharness.html:
  â”” PASS [expected FAIL] StyleSheet interface: sheet must inherit property "ownerNode" with the proper type
```

**Comment by jdm** [comment]
More concretely, those are test results found under `tests/wpt/metadata-layout-2020/`.

**Comment by ghostd** [comment]
I updated the 2020 ini files and rebased onto master

**Pull request merged**